feat(docs): add hash custom instructions and rearange guides (#1400)
flare-foundation/developer-hub: docs/smart-accounts/3-custom-instruction.mdx

slug: custom-instruction
title: Custom Instruction
authors: [nikerzetic]

> description: Performing custom function calls in the Flare Smart Accounts by committing to a PackedUserOperation hash on XRPL and delivering the bytes via an executor (0xFE).

Reviewer: Remove from the description complicated things `PackedUserOperation` and `0xFE`, rewrite it so that users can glance and understand

> ## User operation payload

Reviewer: ```suggestion
## User Operation Payload
```

> A custom instruction has two layers: the outer EIP-4337 [`PackedUserOperation`](https://eips.ethereum.org/EIPS/eip-4337#useroperation) that the XRPL memo commits to, and the inner [`executeUserOp(Call[])`](/smart-accounts/reference/IPersonalAccount#executeuserop) that the personal account runs once the controller dispatches it.

Reviewer: ```suggestion
A custom instruction has two layers: the outer [EIP-4337](https://eips.ethereum.org/EIPS/eip-4337) [`PackedUserOperation`](https://eips.ethereum.org/EIPS/eip-4337#useroperation) that the XRPL memo commits to, and the inner [`executeUserOp(Call[])`](/smart-accounts/reference/IPersonalAccount#executeuserop) that the personal account runs once the controller dispatches it.
```

Only three fields from the [`PackedUserOperation`](https://eips.ethereum.org/EIPS/eip-4337#useroperation) struct are required for Flare Smart Accounts:

- `sender` **must** equal the address of the personal account derived from the XRPL sender.
  Use [`getPersonalAccount`](/smart-accounts/reference/IMasterAccountController#getpersonalaccount) to look it up - the address is deterministic, so you can fetch it before the account is even deployed.
- `nonce` **must** equal the personal account's current nonce returned by [`getNonce`](/smart-accounts/reference/IMasterAccountController#getnonce).
  The nonce auto-increments on every successful execution to prevent replay.

> - `callData` is the calldata that the controller invokes on the personal account.

Reviewer: ```suggestion
- `callData` is the data that the controller invokes on the personal account.
```

});
```

The encoded `callData` becomes the `callData` field of the `PackedUserOperation` that the XRPL memo commits to.

> ## Memo layout

Reviewer: ```suggestion
## Memo Layout
```

| ------- | ---------------- | ------------------------------------------------------------------- |
| `0`     | `instructionId`  | `0xFE` - custom instruction                                         |
| `1`     | `walletId`       | One-byte wallet identifier assigned by Flare; `0` if not registered |
| `2-9`   | `executorFeeUBA` | Executor fee in the FAsset's smallest unit, big-endian `uint64`     |
| `10-41` | `userOpHash`     | `keccak256(abi.encode(userOp))` - the 32-byte commitment            |

The memo length is independent of the call batch: a single small call and a 50-call batch both fit in the same 42 bytes, because the user-operation bytes the executor delivers off-chain never touch the XRPL ledger.
This is the main practical advantage over the [Raw Custom Instruction](/smart-accounts/raw-custom-instruction), whose memo carries the entire ABI-encoded `PackedUserOperation` and is therefore subject to the XRPL's 1024-byte memo cap.

The off-chain delivery also makes the call payload **private on XRPL**.
Only the 32-byte commitment is published; the inner `target`, `value`, and `data` of each call only become visible when the executor submits the user operation to Flare.

> ## Three-step protocol

Reviewer: ```suggestion
## Three-step Protocol
```

> The 0xFE flow runs three steps that map onto two independent actors in production.

Reviewer: ```suggestion
The `0xFE` flow runs three steps that map onto two independent actors.
```

The executor bridges the XRPL payment to Flare with a proof from the [Flare Data Connector (FDC)](/fdc/overview), the same attestation system used by the proof-based flow:

```mermaid
sequenceDiagram
    participant User as User (XRPL)
    participant XRPL
    participant Executor
    participant AM as AssetManagerFXRP
    participant MasterAccountController
    participant PersonalAccount

    User->>User: 1. encode UserOp
    User->>User: compute keccak256(userOp)
    User->>XRPL: 2. Payment to direct-minting<br/>address with 42-byte memo<br/>[0xFE][walletId][fee][hash]
    Note over Executor: receives userOp bytes<br/>out-of-band

    Executor->>Executor: 3. fetch FDC XRPPayment proof
    Executor->>AM: 4. executeDirectMintingWithData(proof, data)<br/>{value: sum(call.value)}
    AM->>MasterAccountController: handleMintedFAssets(..., _memoData, _executor, _data)
    MasterAccountController->>MasterAccountController: keccak256(_data) == hash?
    MasterAccountController->>PersonalAccount: call{value}(executeUserOp(calls))
    MasterAccountController-->>User: UserOperationExecuted event
```

> ### Step 1: user side

Reviewer: ```suggestion
### Step 1: User Side
```

The user constructs the `PackedUserOperation` as described in [User operation payload](#user-operation-payload), computes `keccak256` over the ABI encoding, and packs the 42-byte memo from [Memo layout](#memo-layout).

> The user sends an XRPL `Payment` to the FAssets direct-minting address with this memo, and delivers the full `PackedUserOperation` bytes to the executor **out-of-band** (e.g. over an authenticated HTTP API).

Reviewer: ```suggestion
The user sends an XRPL `Payment` to the FAssets [direct minting address](/fassets/reference/IAssetManager#directmintingpaymentaddress) with this memo, and delivers the full `PackedUserOperation` bytes to the executor **out-of-band** (e.g. over an authenticated HTTP API).
```

> The user sends an XRPL `Payment` to the FAssets direct-minting address with this memo, and delivers the full `PackedUserOperation` bytes to the executor **out-of-band** (e.g. over an authenticated HTTP API).

Reviewer: What does it mean `out-of-band`? Off-chain?

> The bytes never appear on the XRPL ledger.

Reviewer: This is quite confusing, what bytes? Maybe can get rid of this sentence at all.

> ### Step 2: executor side

Reviewer: ```suggestion
### Step 2: Executor Side
```

> The executor takes the XRPL transaction hash, requests an [`IXRPPayment` attestation](/fdc/attestation-types/payment) from the [Flare Data Connector](/fdc/overview), and calls `executeDirectMintingWithData` on `AssetManagerFXRP` (see the [FAssets direct minting page](/fassets/direct-minting)):

Reviewer: `IXRPPayment` or `XRPPayment` ?

```solidity
function executeDirectMintingWithData(
    IXRPPayment.Proof calldata _payment,
    bytes calldata _data
) external payable;
```

- `_payment` is the FDC proof of the XRPL `Payment`.

> - `_data` is the ABI-encoded `PackedUserOperation` that was delivered out-of-band.

Reviewer: `out-of-band` sounds very funky (pun intented)

- `msg.value` **must equal the sum of `call.value` across the user operation**.

>   `AssetManagerFXRP` forwards this value into `MasterAccountController.handleMintedFAssets`, which forwards it again into the personal account's `executeUserOp` so the inner calls can attach native value.

Reviewer: ```suggestion
The `AssetManagerFXRP` forwards this value to the `MasterAccountController` function `handleMintedFAssets`, which forwards it to the personal account's `executeUserOp` function so that the inner calls can attach the native value.
```

The `executeDirectMintingWithData` function is **only valid for smart-account targets** - calling it for a non-smart-account direct mint reverts.

> ### Step 3: confirmation

Reviewer: ```suggestion
### Step 3: Confirmation
```

The `MasterAccountController` verifies on-chain that `keccak256(_data) == userOpHash` from the memo.
If it matches, it decodes `_data` as a `PackedUserOperation`, validates `sender` and `nonce`, executes `executeUserOp` on the personal account, and emits [`UserOperationExecuted`](/smart-accounts/reference/IMasterAccountController#useroperationexecuted) - **all inside the executor's transaction**.
This is the key difference from the proof-based dispatch: there is no separate cross-chain wait, because the executor's call already executed the user operation by the time it returns.

> ## Hash mismatch

Reviewer: ```suggestion
## Hash Mismatch
```

If the bytes the executor submits do not hash to the commitment in the memo, `handleMintedFAssets` reverts with `CustomInstructionHashMismatch(expected, actual)`.

> The FAsset transfer is performed before the memo is decoded, however, so even on a mismatch the FXRP credited by direct minting remains in the personal account, and the user can recover by issuing a fresh user operation (see [Failure Handling](#failure-handling)).

Reviewer: ```suggestion
The FAsset transfer is performed before the memo is decoded, however, so even on a mismatch the FXRP credited by direct minting remains in the personal account, and the user can recover by issuing a new user operation (see [Failure Handling](#failure-handling)).
```

> ## Call value accounting

Reviewer: ```suggestion
## Call Value Accounting
```

Whatever native value the executor attaches to `executeDirectMintingWithData` is forwarded all the way to `executeUserOp` (`AssetManagerFXRP -> MasterAccountController.handleMintedFAssets -> PersonalAccount.executeUserOp`).

> The executor must therefore compute the total native value to attach as the sum of `call.value` across the user operation it received out-of-band.

Reviewer: Out of band ?

The user-side helper in the [TypeScript guide](/smart-accounts/guides/typescript-viem/custom-instruction-ts) returns this value alongside the XRPL transaction hash, so the executor does not have to recompute it from scratch.

> ## Replay protection

Reviewer: ```suggestion
## Replay Protection
```
